include/albatross/src/core/parameters.hpp

```diff
@@ -202,6 +202,107 @@ inline bool params_are_valid(const ParameterStore &params) {
   return true;
 }
 
+inline bool modify_param(const ParameterKey &name,
+                         const std::function<void(Parameter *)> &operation,
+                         ParameterStore *params, bool assert_exists = true) {
+  const auto it = params->find(name);
+  if (it != params->end()) {
+    operation(&it->second);
+    return true;
+  }
+
+  if (assert_exists) {
+    std::cerr << "Error: Parameter `" << name << "` not found";
+    ALBATROSS_ASSERT(false && "Attempt to modify a param which doesn't exist");
+  }
+  return false;
+}
+
+inline void set_param(const ParameterKey &name, const Parameter &param,
```

Comment:
You could imagine avoiding these 3 wrappers by returning something like
`std::optional<std::reference_wrapper<Parameter>>` or `Parameter*`.

Reply:
Can you explain?  We actually don't have `optional` in albatross, it's not part of c++14 and while we could pull in a non standard implementation I've trying to limit dependencies ... but we could make it happen if we need to.

Reply:
What I mean is that if you return some sort of reference to the `Parameter` in question, the caller can get this reference and modify the parameter, the value, the key etc. as needed, without the need for multiple similar functions in the API (kind of like `std::map<Key, T>::at()` returns `T&`).

Then the question becomes how to return such a reference when the right parameter exists but also return something meaningful when it does not.  The two types I mentioned are ways to do that; the first has the drawbacks you mentioned, and the second has the drawback that the compiler doesn't really encourage you to check that it's non-`nullptr` before you dereference it.

Reply:
Ah yeah, I think I actually tried that first and you've identified some of the pain points which made me change course.

The API involved:
```
Parameter *get_param_pointer(const ParameterKey &key)
```
instead of:
```
void set_param(const ParameterKey &key, const Parameter &param);
```

What I found was that this greatly simplified all the `if_exists` logic and was almost certainly more computationally efficient ... but the deal breaker was that once a class returns a pointer to a parameter it has no clue if that parameter has been changed and that's critical to some of our applications in which we have parameters which require post processing.  For example, a class which requires a grid, you might parameterize it with a `grid_origin` and `grid_spacing` parameters, but internally you'd store a `std::vector<double> grid_points`.  With the `set_param` logic you give the implementing class the opportunity to update the grid points when a parameter has changed, with the `get_param_pointer` approach you have no clue when (or if) the parameter has changed.

Here's the branch if you're curious: https://github.com/swift-nav/albatross/compare/9479ecd2a7608530d5b0e3ed8e71c8b6c2237582...akleeman/param_free_functions_get_ptr

Reply:
The need to trigger further processing on changes seems like a great reason to stick with the current API.  Thanks for clarifying.